src/evaluators/scatter/PHAL_ScatterSideEqnResidual_Def.hpp

```diff
@@ -293,19 +322,26 @@ doPostEvaluate(typename Traits::EvalData workset)
   // Set J=identity outside of the sideset, so it's not singular
   auto Jac = workset.Jac;
   const int numCellNodes = node_dof_mgr->getGIDFieldOffsets(0).size();
+  const auto& ss_eqn_lids = this->ss_eqns_dofs_lids.at(workset.wsIndex);
+  const RealType j_coeff = workset.j_coeff;
+  Teuchos::Array<ST> vals;
+  Teuchos::Array<LO> cols;
   for (size_t icell=0; icell<workset.numCells; ++icell) {
     const auto elem_LID = elem_lids(icell);
     const auto dof_lids = Kokkos::subview(elem_dof_lids,elem_LID,ALL);
 
-    const auto& node_gids = node_dof_mgr->getElementGIDs(elem_LID);
-
-    for (int inode=0; inode<numCellNodes; ++inode) {
-      const auto node = node_gids[inode];
-      if (this->ss_nodes_gids.count(node)==0) {
-        for (int eq = 0; eq < this->numFields; eq++) {
-          const auto& offsets = dof_mgr->getGIDFieldOffsets(eq+this->offset);
-          const LO lrow = dof_lids(offsets[inode]);
-          Albany::setLocalRowValue(Jac,lrow,lrow, 1.0);
+    for (int eq = 0; eq < this->numFields; eq++) {
+      const auto& offsets = dof_mgr->getGIDFieldOffsets(eq+this->offset);
+      for (auto o : offsets) {
+        auto lrow = dof_lids[o];
+        if (ss_eqn_lids.count(lrow)==0) {
+          // LID not on the sideset. Diagonalize
+
+          // Extract the row, zero it out, then put j_coeff on diagonal
+          Albany::getLocalRowValues(Jac, lrow, cols, vals);
+          for (auto& val : vals) { val = 0.0; }
```

Comment:
Should this line be a Kokkos::deep_copy? Probably doesn't matter for now since this evaluator runs on host.

Reply:
Well, `vals` is a `Teuchos::Array<ST>`, so a few things will have to be changed to get this to a device friendly state...

Reply:
Ah, I confused it with an `ArrayView`, nevermind.